pkg/auth/providers/saml/saml_client.go

```diff
@@ -0,0 +1,385 @@
+func (s *Provider) HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml.Assertion) {
+	var groupPrincipals []v3.Principal
+	var userPrincipal v3.Principal
+
+	if relayState := r.Form.Get("RelayState"); relayState != "" {
+		// delete the cookie
+		s.ClientState.DeleteState(w, r, relayState)
+	}
+
+	samlData := make(map[string][]string)
+
+	for _, attributeStatement := range assertion.AttributeStatements {
+		for _, attr := range attributeStatement.Attributes {
+			attrName := attr.FriendlyName
+			if attrName == "" {
+				attrName = attr.Name
+			}
+			for _, value := range attr.Values {
+				samlData[attrName] = append(samlData[attrName], value.Value)
+			}
+		}
+	}
+
+	config, err := s.getSamlConfig()
+	if err != nil {
+		writeError(w, 500, "SAML: Error getting saml config %v", err, "Server error while authenticating")
+		return
+	}
+
+	userPrincipal, groupPrincipals = s.getSamlPrincipals(config, samlData)
+	allowedPrincipals := config.AllowedPrincipalIDs
+
+	allowed, err := s.UserMGR.CheckAccess(config.AccessMode, allowedPrincipals, userPrincipal, groupPrincipals)
+	if err != nil {
+		writeError(w, 500, "SAML: Error during login while checking access %v", err, "Server error while authenticating")
+		return
+	}
+	if !allowed {
+		writeError(w, 403, "SAML: User does not have access %v", err, "User does not have access")
+		return
+	}
+
+	if !config.Enabled {
+		header := make(map[string][]string)
+		header["Impersonate-User"] = []string{config.Annotations["userID"]}
+		newReq := &types.APIContext{
+			Request: &http.Request{
+				Header: header,
+			},
+		}
+		user, err := s.UserMGR.SetPrincipalOnCurrentUser(newReq, userPrincipal)
+		if err != nil {
+			writeError(w, 500, "SAML: Error setting principal on current user %v", err, "Server error while authenticating")
+			return
+		}
+
+		config.Enabled = true
+		err = s.saveSamlConfig(config)
+		if err != nil {
+			writeError(w, 500, "SAML: Error saving SAML config %v", err, "Server error while authenticating")
+			return
+		}
+
+		isSecure := false
+		if r.URL.Scheme == "https" {
+			isSecure = true
+		}
+		setRancherToken(w, r, s.TokenMGR, user.Name, userPrincipal, groupPrincipals, isSecure)
+		return
+	}
+
+	displayName := userPrincipal.DisplayName
+	if displayName == "" {
+		displayName = userPrincipal.LoginName
+	}
+	user, err := s.UserMGR.EnsureUser(userPrincipal.Name, displayName)
+	if err != nil {
+		writeError(w, 403, "SAML: User does not have access %v", err, "User does not have access")
+		return
+	}
+
+	setRancherToken(w, r, s.TokenMGR, user.Name, userPrincipal, groupPrincipals, true)
+	return
+}
+
+func setRancherToken(w http.ResponseWriter, r *http.Request, tokenMGR *tokens.Manager, userID string, userPrincipal v3.Principal,
+	groupPrincipals []v3.Principal, isSecure bool) {
+	rToken, err := tokenMGR.NewLoginToken(userID, userPrincipal, groupPrincipals, "", 0, "")
+	if err != nil {
```

Comment:
dont have setRancherToken writeError. Have it return an error and the caller decide what to do with it (write it)